src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -362,5 +411,133 @@ private void Cancel()
         {
             InternalTokenSource.Cancel();
         }
+
+        static void ThrowConcurrentReadsNotSupported()
+        {
+            throw new InvalidOperationException($"Concurrent reads are not supported; await the {nameof(ValueTask<ReadResult>)} before starting next read.");
+        }
+
+        private class AsyncReader : IValueTaskSource<ReadResult>
+        {
+            private readonly StreamPipeReader _reader;
+            private readonly Action _onCompletedAction;
+
+            private ManualResetValueTaskSourceCore<ReadResult> _mrvts;
+
+            private ValueTaskAwaiter<int> _readAwaiter;
+            private CancellationTokenSource _tokenSource;
```

Comment:
Access the token source from the _reader.

Reply:
Done